File: personal-document-renewal-vault-poc/backend/adapters.py
"""Optional PostgreSQL, Redis, and MinIO infrastructure adapters."""

from __future__ import annotations

import json
import os
from io import BytesIO
from typing import Any


class Infrastructure:
    def __init__(self) -> None:
        self.mode = "memory"
        self.postgres: Any = None
        self.redis: Any = None
        self.minio: Any = None

    def connect(self) -> "Infrastructure":
        database_url = os.getenv("VAULT_DATABASE_URL", "memory://")
        redis_url = os.getenv("VAULT_REDIS_URL", "memory://")
        minio_endpoint = os.getenv("VAULT_MINIO_ENDPOINT", "memory://")
        if "memory://" in {database_url, redis_url, minio_endpoint}:
            return self
        try:
            import psycopg
            import redis
            from minio import Minio

            self.postgres = psycopg.connect(database_url, autocommit=True)
            self.postgres.execute(
                """
                CREATE TABLE IF NOT EXISTS document_vault_snapshots (
                    id TEXT PRIMARY KEY,
                    state JSONB NOT NULL,
                    updated_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
                )
                """
            )
            self.redis = redis.Redis.from_url(redis_url, decode_responses=True)
            self.redis.ping()
            self.minio = Minio(
                minio_endpoint,
                access_key=os.getenv("VAULT_MINIO_ACCESS_KEY", "minioadmin"),
                secret_key=os.getenv("VAULT_MINIO_SECRET_KEY", "minioadmin"),
                secure=os.getenv("VAULT_MINIO_SECURE", "false").lower() == "true",
            )
            bucket = os.getenv("VAULT_MINIO_BUCKET", "personal-documents")
            if not self.minio.bucket_exists(bucket):
                self.minio.make_bucket(bucket)
            self.mode = "postgres+redis+minio"
        except Exception as exc:
            print(f"Infrastructure unavailable ({exc}); using memory mode")
            self.close()
        return self

    def load(self) -> dict[str, Any] | None:
        if self.mode == "memory":
            return None
        row = self.postgres.execute(
            "SELECT state FROM document_vault_snapshots WHERE id = %s", ("demo",)
        ).fetchone()
        return row[0] if row else None

    def save(self, state: dict[str, Any]) -> None:
        if self.mode == "memory":
            return
        payload = json.dumps(state)
        self.postgres.execute(
            """
            INSERT INTO document_vault_snapshots (id, state, updated_at)
            VALUES (%s, %s::jsonb, NOW())
            ON CONFLICT (id) DO UPDATE SET state = EXCLUDED.state, updated_at = NOW()
            """,
            ("demo", payload),
        )
        self.redis.setex("document-vault:snapshot", 300, payload)

    def put_object(self, object_key: str, content: bytes, content_type: str) -> None:
        if self.mode == "memory":
            return
        bucket = os.getenv("VAULT_MINIO_BUCKET", "personal-documents")
        self.minio.put_object(
            bucket,
            object_key,
            BytesIO(content),
            length=len(content),
            content_type=content_type,
        )

    def get_object(self, object_key: str) -> bytes | None:
        if self.mode == "memory":
            return None
        bucket = os.getenv("VAULT_MINIO_BUCKET", "personal-documents")
        response = self.minio.get_object(bucket, object_key)
        try:
            return response.read()
        finally:
            response.close()
            response.release_conn()

    def mirror_jobs(self, jobs: list[dict[str, Any]]) -> None:
        if self.mode == "memory":
            return
        key = "document-vault:jobs"
        self.redis.delete(key)
        for job in jobs:
            if job["status"] != "COMPLETED":
                self.redis.rpush(key, json.dumps(job))

    def close(self) -> None:
        for connection in (self.redis, self.postgres):
            try:
                if connection:
                    connection.close()
            except Exception:
                pass
        self.postgres = self.redis = self.minio = None
        self.mode = "memory"

File: home-utility-usage-monitor-poc/backend/src/adapters.ts
import { Kafka, type Consumer, type Producer } from "kafkajs";
import pg from "pg";
import { createClient, type RedisClientType } from "redis";
import type { Reading } from "./core.js";

type KafkaMessage = {
  key: string;
  value: Record<string, unknown>;
};

export class Infrastructure {
  mode = "memory";
  kafkaMode = "memory";
  postgresMode = "memory";
  redisMode = "memory";
  messages: KafkaMessage[] = [];
  producer: Producer | null = null;
  consumer: Consumer | null = null;
  pgPool: pg.Pool | null = null;
  redis: RedisClientType | null = null;
  private snapshot: Record<string, unknown> | null = null;

  async connect(): Promise<this> {
    await Promise.all([this.connectPostgres(), this.connectRedis(), this.connectKafka()]);
    this.mode = [this.postgresMode, this.redisMode, this.kafkaMode].join("+");
    return this;
  }

  async connectPostgres(): Promise<void> {
    const databaseUrl = process.env.UTILITY_DATABASE_URL || "memory://";
    if (databaseUrl === "memory://") return;
    try {
      this.pgPool = new pg.Pool({ connectionString: databaseUrl });
      await this.pgPool.query(`
        CREATE TABLE IF NOT EXISTS utility_monitor_snapshots (
          id TEXT PRIMARY KEY,
          state JSONB NOT NULL,
          updated_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
        )
      `);
      await this.pgPool.query(`
        CREATE TABLE IF NOT EXISTS utility_meter_readings (
          event_key TEXT PRIMARY KEY,
          meter_id TEXT NOT NULL,
          measured_at TIMESTAMPTZ NOT NULL,
          value DOUBLE PRECISION NOT NULL,
          payload JSONB NOT NULL
        )
      `);
      await this.pgPool.query(`CREATE INDEX IF NOT EXISTS utility_meter_readings_meter_time ON utility_meter_readings (meter_id, measured_at)`);
      try {
        await this.pgPool.query("CREATE EXTENSION IF NOT EXISTS timescaledb");
        await this.pgPool.query("SELECT create_hypertable('utility_meter_readings', 'measured_at', if_not_exists => TRUE)");
        this.postgresMode = "postgres+timescale";
      } catch {
        this.postgresMode = "postgres";
      }
    } catch (error) {
      console.warn(`PostgreSQL unavailable (${(error as Error).message}); using memory persistence`);
      await this.closePostgres();
    }
  }

  async connectRedis(): Promise<void> {
    const redisUrl = process.env.UTILITY_REDIS_URL || "memory://";
    if (redisUrl === "memory://") return;
    try {
      this.redis = createClient({ url: redisUrl });
      await this.redis.connect();
      this.redisMode = "redis";
    } catch (error) {
      console.warn(`Redis unavailable (${(error as Error).message}); using memory alert state`);
      await this.closeRedis();
    }
  }

  async connectKafka(): Promise<void> {
    const brokers = (process.env.UTILITY_KAFKA_BROKERS || "memory://").split(",").filter(Boolean);
    if (brokers.includes("memory://")) return;
    try {
      const kafka = new Kafka({ clientId: "home-utility-monitor", brokers });
      this.producer = kafka.producer();
      this.consumer = kafka.consumer({ groupId: process.env.UTILITY_KAFKA_GROUP || "utility-monitor-poc" });
      await this.producer.connect();
      await this.consumer.connect();
      this.kafkaMode = "kafka";
    } catch (error) {
      console.warn(`Kafka unavailable (${(error as Error).message}); using memory broker`);
      await this.closeKafka();
    }
  }

  async publishReading(topic: string, key: string, value: Record<string, unknown>): Promise<void> {
    if (this.producer) {
      await this.producer.send({ topic, messages: [{ key, value: JSON.stringify(value) }] });
      return;
    }
    this.messages.push({ key, value: structuredClone(value) });
  }

  async startKafkaConsumer(topic: string, handler: (message: Record<string, unknown>) => Promise<void>): Promise<boolean> {
    if (!this.consumer) return false;
    await this.consumer.subscribe({ topic, fromBeginning: false });
    await this.consumer.run({
      eachMessage: async ({ message }) => {
        if (!message.value) return;
        const payload = JSON.parse(message.value.toString()) as Record<string, unknown>;
        await handler(payload);
      }
    });
    return true;
  }

  async consumeMemoryMessages(handler: (message: Record<string, unknown>) => Promise<void>): Promise<{ processed: number }> {
    let processed = 0;
    while (this.messages.length) {
      const message = this.messages.shift();
      if (!message) break;
      await handler(message.value);
      processed += 1;
    }
    return { processed };
  }

  async save(state: Record<string, unknown>): Promise<void> {
    this.snapshot = structuredClone(state);
    if (this.pgPool) {
      await this.pgPool.query(
        `
        INSERT INTO utility_monitor_snapshots (id, state, updated_at)
        VALUES ($1, $2::jsonb, NOW())
        ON CONFLICT (id) DO UPDATE SET state = EXCLUDED.state, updated_at = NOW()
        `,
        ["demo", JSON.stringify(state)]
      );
    }
    if (this.redis) await this.redis.setEx("utility-monitor:snapshot", 300, JSON.stringify(state));
  }

  async load(): Promise<Record<string, unknown> | null> {
    if (this.pgPool) {
      const result = await this.pgPool.query("SELECT state FROM utility_monitor_snapshots WHERE id = $1", ["demo"]);
      return result.rows[0]?.state || null;
    }
    return this.snapshot;
  }

  async appendReading(reading: Reading): Promise<void> {
    if (this.pgPool) {
      await this.pgPool.query(
        `
        INSERT INTO utility_meter_readings (event_key, meter_id, measured_at, value, payload)
        VALUES ($1, $2, $3, $4, $5::jsonb)
        ON CONFLICT (event_key) DO NOTHING
        `,
        [reading.eventKey, reading.meterId, reading.measuredAt, reading.value, JSON.stringify(reading)]
      );
    }
  }

  async mirrorJobs(jobs: Array<Record<string, unknown>>): Promise<void> {
    if (!this.redis) return;
    await this.redis.del("utility-monitor:jobs");
    for (const job of jobs) {
      if (job.status === "READY" || job.status === "RETRY") {
        await this.redis.rPush("utility-monitor:jobs", JSON.stringify(job));
      }
    }
  }

  async closePostgres(): Promise<void> {
    if (this.pgPool) await this.pgPool.end().catch(() => undefined);
    this.pgPool = null;
    this.postgresMode = "memory";
  }

  async closeRedis(): Promise<void> {
    if (this.redis?.isOpen) await this.redis.quit().catch(() => undefined);
    this.redis = null;
    this.redisMode = "memory";
  }

  async closeKafka(): Promise<void> {
    if (this.consumer) await this.consumer.disconnect().catch(() => undefined);
    if (this.producer) await this.producer.disconnect().catch(() => undefined);
    this.consumer = null;
    this.producer = null;
    this.kafkaMode = "memory";
  }

  async close(): Promise<void> {
    await Promise.all([this.closeKafka(), this.closeRedis(), this.closePostgres()]);
  }
}
